File: src/main/java/komu/blunt/asm/Linkage.java
package komu.blunt.asm;

import static com.google.common.base.Preconditions.checkNotNull;

public final class Linkage {
    
    public static final Linkage NEXT = new Linkage("next", null);
    public static final Linkage RETURN = new Linkage("return", null);

    private final String name;
    public final Label label;

    private Linkage(String name, Label label) {
        this.name = checkNotNull(name);
        this.label = label;
    }

    public static Linkage jump(Label label) {
        return new Linkage("jump", checkNotNull(label));
    }

    @Override
    public String toString() {
        return label != null ? name + " " + label : name;
    }
}

File: src/main/kotlin/komu/blunt/asm/Linkage.kt
package komu.blunt.asm

sealed class Linkage() {

    object Next : Linkage() {
        override fun toString() = "next"
    }

    object Return : Linkage() {
        override fun toString() = "return"
    }

    class Jump(val label: Label) : Linkage() {
        override fun toString() = "jump $label"
    }
}
